### PR TITLE
Implement gradient for QR decomposition

### LINKED ISSUE #1099: Implement gradient for QR decomposition
### Description

QR is one of the few remaining linalg ops that is missing a gradient. JAX code for the jvp is [here](https://github.com/jax-ml/jax/blob/73fa0f48cb0081fc69cb82dffe6ceb5433cdc446/jax/_src/lax/linalg.py#L1954), whic also includes [this derivation](https://j-towns.github.io/papers/qr-derivative.pdf). [This paper ](https://www.tandfonline.com/doi/abs/10.1080/10556788.2011.610454) also claims to derive the gradients for QR, but I find it unreadable.

Relatedly but perhaps worthy of a separate issue, [this paper ](https://arxiv.org/abs/1710.08717) derives gradients for the LQ decomposition, $$A = LQ$$, where $L$ is lower triangular and $Q$ is orthonormal ($$Q^TQ=I$$.)  Compare this to QR, which gives you $$A = QR$$, where $$Q$$ is again orthonormal, but $$R$$ is upper triangular, and you see why I mention it in this issue. It wouldn't be hard to offer LQ as well.

### PR DESCRIPTION
**Pull request by educhesne**
## Description
Implement the support for gradient in the QR decomposition. It is only defined when R is a square matrix.

## Related Issue
- [x] Closes #1099 

## Checklist
- [x] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [x] Included tests that prove the fix is effective or that the new feature works
- [x] Added necessary documentation (docstrings and/or example notebooks)
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Type of change

- [x] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):

----
📚 Documentation preview 📚: https://pytensor--1303.org.readthedocs.build/en/1303/

### CODE REVIEW

File: pytensor/tensor/nlinalg.py [hunk 1]
```diff
@@ -512,6 +512,78 @@ def perform(self, node, inputs, outputs):
...
+            )
+
+        elif self.mode == "r":
+            # We need all the components of the QR to compute the gradient of A even if we only
+            # use the upper triangular component in the cost function.
+            Q, R = qr(A, mode="reduced")
+            dR = cast(ptb.TensorVariable, output_grads[0])
+            R_dRt = R @ _H(dR)
+            Rinvt = _H(inv(R))
+            A_bar = Q @ ((ptb.tril(R_dRt - _H(R_dRt), k=-1)) @ Rinvt + dR)
```
Comment (jessegrabowski) [inline pytensor/tensor/nlinalg.py:554]: Use a solve here instead of explicitly inverting `R`. You can use `solve_triangular` to exploit the structure of R, and set `trans=1` instead of inverting R at all.

File: pytensor/tensor/nlinalg.py [hunk 2]
```diff
@@ -512,6 +512,78 @@ def perform(self, node, inputs, outputs):
...
+                is_disconnected, output_grads, [Q, R], strict=True
+            ):
+                if disconnected:
+                    new_output_grads.append(output.zeros_like())
+                else:
+                    new_output_grads.append(output_grad)
+
+            (dQ, dR) = (cast(ptb.TensorVariable, x) for x in new_output_grads)
+
+            Rinvt = _H(inv(R))
```
Comment (jessegrabowski) [inline pytensor/tensor/nlinalg.py:578]: Same as above, avoid ever using `inv` and use a `solve_triangular` instead

File: tests/tensor/test_nlinalg.py [hunk 3]
```diff
@@ -152,6 +152,18 @@ def test_qr_modes():
         assert "name 'complete' is not defined" in str(e)
 
 
+@pytest.mark.parametrize("shape", [(3, 3), (6, 3)], ids=["shape=(3, 3)", "shape=(6,3)"])
+@pytest.mark.parametrize("output", [0, 1], ids=["Q", "R"])
+def test_qr_grad(shape, output):
```
Comment (jessegrabowski) [inline tests/tensor/test_nlinalg.py:157]: parameterize on `mode` as well, and test a case where both Q and R are implicated in the loss function.

**Review by jessegrabowski — CHANGES_REQUESTED** [review]
This is really great! The most important change that we need here is to use `solve_triangular` instead of inverse. 

If you're willing, I would also like to move the QR function from `nlinalg` to `slinalg`, and use the scipy signature in the perform method. If you're not that's fine, and we can do that in a future PR.

**Comment by educhesne** [comment]
ok thanks ! I didn't use the trans argument of solve_triangular though; would that simplify the expression ? (it works anyway)

File: pytensor/tensor/nlinalg.py [hunk 4]
```diff
@@ -512,6 +512,78 @@ def perform(self, node, inputs, outputs):
...
+        .. [2] Sebastian F. Walter , Lutz Lehmann & René Lamour. "On evaluating higher-order derivatives
+        of the QR decomposition of tall matrices with full column rank in forward and reverse mode algorithmic differentiation",
+        Optimization Methods and Software, 27:2, 391-403, DOI: 10.1080/10556788.2011.610454
+        """
+
+        (A,) = (cast(ptb.TensorVariable, x) for x in inputs)
+        *_, m, n = A.type.shape
+
+        def _H(x: ptb.TensorVariable):
+            return x.conj().T
```
[the remark this answers is not in the document]
Comment (ricardoV94) [reply pytensor/tensor/nlinalg.py]: That's fine the gradient should only see 2D stuff

[the remark this answers is not in the document]
Comment (educhesne) [reply pytensor/tensor/nlinalg.py]: I already updated it with `mT` anyway...

[the remark this answers is not in the document]
Comment (jessegrabowski) [reply pytensor/tensor/nlinalg.py]: mT is more correct :)

**Comment by jessegrabowski** [comment]
No worries about the `trans` argument, I think I misunderstood what was needed.

This is really great! Thanks for your contribution. It looks nice to me. Going to ask @ricardoV94 for a final sign off then merge.

**Review by jessegrabowski — APPROVED** [review]

File: pytensor/tensor/nlinalg.py [hunk 5]
```diff
@@ -512,6 +512,80 @@ def perform(self, node, inputs, outputs):
...
+        def _H(x: ptb.TensorVariable):
+            return x.conj().mT
+
+        def _copyutl(x: ptb.TensorVariable):
+            return ptb.triu(x, k=0) + _H(ptb.triu(x, k=1))
+
+        if self.mode == "raw" or (self.mode == "complete" and m != n):
+            raise NotImplementedError("Gradient of qr not implemented")
+
+        elif m < n:
```
Comment (ricardoV94) [inline pytensor/tensor/nlinalg.py:544]: gradient should work for non-static shapes, `A = matrix("A", shape=(None, None))`. If you need the shape check add it symbolically with `assert_op`

File: pytensor/tensor/nlinalg.py [hunk 6]
```diff
@@ -512,6 +512,80 @@ def perform(self, node, inputs, outputs):
...
+        *_, m, n = A.type.shape
+
+        def _H(x: ptb.TensorVariable):
+            return x.conj().mT
+
+        def _copyutl(x: ptb.TensorVariable):
+            return ptb.triu(x, k=0) + _H(ptb.triu(x, k=1))
+
+        if self.mode == "raw" or (self.mode == "complete" and m != n):
+            raise NotImplementedError("Gradient of qr not implemented")
```
Comment (ricardoV94) [inline pytensor/tensor/nlinalg.py:542]: Mention what exactly makes it non-implemented

**Review by ricardoV94 — CHANGES_REQUESTED** [review]
This is awesome!

Just some small change requests

File: pytensor/tensor/nlinalg.py (hunk 5 above)
Comment (educhesne) [reply pytensor/tensor/nlinalg.py]: I replaced the shape checks with `Assert` on the output shape. Let me know if that's what you had in mind, I'm not very familiar with symbolic shapes

**Review by jessegrabowski — APPROVED** [review]
lgtm

**Comment by educhesne** [comment]
I'm really sorry, wait a bit before merging, I may have mixed conjugate transpose and transpose. The tests are only done with real matrix, so that wouldn't be caught.

**Comment by jessegrabowski** [comment]
Can you try adding a complex test? You can follow the template [here](https://github.com/pymc-devs/pytensor/blob/8a7356ce7e0bf1a567921ad0c01d401eadba643b/tests/link/numba/test_slinalg.py#L204). Those tests don't work because numba is fussy. I think this case won't work either because we don''t support complex gradients, but you can still add the tests and xfail them (it's aspirational that way!)

**Comment by educhesne** [comment]
Ok, done.
I've rechecked the conjugate/transpose thing and I am more confident it is correct now, sorry I have overreacted...

**Comment by jessegrabowski** [comment]
Did you check if they actually do fail before you xfailed them? I'm not 100% sure on that

**Comment by educhesne** [comment]
yes I did, `verify_grad` doesn't accept [complex dtype](https://github.com/pymc-devs/pytensor/blob/8a7356ce7e0bf1a567921ad0c01d401eadba643b/pytensor/gradient.py#L1877C36-L1877C38)

**Review by ricardoV94 — APPROVED** [review]
Looks great! This is gold

**Comment by educhesne** [comment]
Actually @jessegrabowski even `pytensor.gradient.grad` doesn't accept complex dtypes.
I found old discussions from the Theano era about it, but nothing else (in particular nothing about choosing the convention for complex gradients, between the jax/holomorphic compatible convention vs the torch/gradient descent compatible one)
Do you know if it is something planned or considered ?

**Comment by jessegrabowski** [comment]
There's been some vague conversations about it, but nothing concrete has been planned. It needs a dev to come along and take up the banner to make it happen

**Comment by educhesne** [comment]
Ok thanks. I could help, but being new to pytensor dev I don't know about being the flag-bearer...

**Comment by jessegrabowski** [comment]
If you're excited about it, or you need it for you work, you should definitely hack around and see if you can put a prototype, and open a draft PR. If you build it, they will come :)

**Comment by educhesne** [comment]
Ok, I'll see... :)
(I don't need it, but that looks interesting...)

**Comment by jessegrabowski** [comment]
Are you a student interested in doing GSoC? We could make a project like that if you are and it's something you're interested in. Or we can brainstorm something else.

**Comment by educhesne** [comment]
Thanks but I am not a student :-D

**Comment by jessegrabowski** [comment]
Ah ok! Sorry I just assumed. We get a lot of activity around this time of year from students looking to participate in GSoC. Anyway still happy to brainstorm on projects/features you'd like to see in pytensor (if you're looking for stuff to do) :)

**Comment by ricardoV94** [comment]
> Thanks but I am not a student :-D

fyi GSOC is no longer restricted to students since last year, it now also accommodates people who are beginning to contribute to OSS in general
